### PR TITLE
Create a friendlier set of access methods to states on transactions

### PR DESCRIPTION
**Pull request by mnesbit**
This puts various mapping and filtering operation onto BaseTransaction and LedgerTransaction to aid CorDapp developers.

### CODE REVIEW

**Comment by joeldudleyr3** [comment]
This looks good. But why the split of methods between BaseTransaction and LedgerTransaction?

**Comment by mnesbit** [comment]
The two inherited classes are WireTransaction and LedgerTransaction. You can only do input queries on LedgerTransaction where they have been resolve. However, loads of our code and tests just looks at outouts and has just a WireTransaction. I was able to port many more places once I pushed it down a level.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 1]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
     override fun hashCode() = Objects.hash(notary, mustSign, type, timeWindow)
 
     override fun toString(): String = "${javaClass.simpleName}(id=$id)"
+
+    /**
+     * Returns a [StateAndRef] for the given output index.
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun <T : ContractState> outRef(index: Int): StateAndRef<T> = StateAndRef(outputs[index] as TransactionState<T>, StateRef(id, index))
+
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: Extra blank line

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 2]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
...
+    /**
+     * Helper to simplify getting all output states of a particular class, interface, or base class.
+     * @param clazz The class type used for filtering via an isAssignable check.
+     * Clazz must be an extension of [ContractState].
+     * @return the possibly empty list of output states matching the clazz restriction.
+     */
+    fun <T : ContractState> outputsOfType(clazz: Class<T>): List<T> {
+        @Suppress("UNCHECKED_CAST")
+        return outputs.filter { clazz.isAssignableFrom(it.data.javaClass) }.map { it.data as T }
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: You can use `filterIsInstance` which takes in a Class

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 3]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
...
+     * Helper to simplify getting an indexed output.
+     * @param index the position of the item in the output.
+     * @return The ContractState at the requested index
+     */
+    fun getOutput(index: Int): ContractState = outputs[index].data
+
+    /**
+     * Helper to simplify getting all output states of a particular class, interface, or base class.
+     * @param clazz The class type used for filtering via an isAssignable check.
+     * Clazz must be an extension of [ContractState].
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: [clazz]

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 4]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
...
+
+    /**
+     * Helper to simplify filtering outputs according to a [Predicate].
+     * @param predicate A filtering function taking a state of type T and returning true if it should be included in the list.
+     * The class filtering is applied before the predicate.
+     * @param clazz The class type used for filtering via an isAssignable check.
+     * Clazz must be an extension of [ContractState].
+     * @return the possibly empty list of output states matching the predicate and clazz restrictions.
+     */
+    fun <T : ContractState> filterOutputs(predicate: Predicate<T>, clazz: Class<T>): List<T> {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: I don't think we should have this. It's easy enough for Kotlin users to filter the output of `outputsOfType` and Java has the streams API.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 5]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
...
+    /**
+     * Helper to simplify finding a single output matching a [Predicate].
+     * @param predicate A filtering function taking a state of type T and returning true if this is the desired item.
+     * The class filtering is applied before the predicate.
+     * @param clazz The class type used for filtering via an isAssignable check.
+     * Clazz must be an extension of [ContractState].
+     * @return the single item matching the predicate.
+     * @throws IllegalArgumentException if no item, or multiple items are found matching the requirements.
+     */
+    fun <T : ContractState> findOutput(predicate: Predicate<T>, clazz: Class<T>): T {
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: likewise

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 6]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
...
+    /**
+     * Helper to simplify getting all output [StateAndRef] items of a particular state class, interface, or base class.
+     * @param clazz The class type used for filtering via an isAssignable check.
+     * Clazz must be an extension of [ContractState].
+     * @return the possibly empty list of output [StateAndRef<T>] states matching the clazz restriction.
+     */
+    fun <T : ContractState> outRefsOfType(clazz: Class<T>): List<StateAndRef<T>> {
+        @Suppress("UNCHECKED_CAST")
+        return outputs.mapIndexed { index, state -> StateAndRef(state, StateRef(id, index)) }
+                .filter { clazz.isAssignableFrom(it.state.data.javaClass) }
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: You have the filter first.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 7]
```diff
@@ -56,4 +58,126 @@ abstract class BaseTransaction(
...
+    fun <T : ContractState> findOutRef(predicate: Predicate<T>, clazz: Class<T>): StateAndRef<T> {
+        return filterOutRefs(predicate, clazz).single()
+    }
+
+}
+
+/**
+ * Kotlin extension methods to take advantage of Kotlin's smart type inference when querying the LedgerTransaction
+ */
+inline fun <reified T : ContractState> BaseTransaction.outputsOfType(): List<T> = this.outputsOfType(T::class.java)
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: You can move these inside `BaseTransaction` - Java code won't see it (just like in FlowLogic).

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt (hunk 4 above)
Comment (joeldudleyr3) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: @shamsasari Could you give an example usage? Whatever I try and use streams, it's incredibly verbose.

**Comment by mnesbit** [comment]
I do understand the question about streams, but actually I am quite adamant that is the wrong choice, because people write the same code over and over and over again. It is ugly, long winded and shows very clearly that we haven't written our API to be used. Giving what are just syntactic sugar vastly helps the user friendliness.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt (hunk 7 above)
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: Didn't know I could do that I will push them inside the classes.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt (hunk 6 above)
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: If I do that I was worried about losing the index, but I will re-arrange.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt (hunk 2 above)
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: OK I will investigate

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt (hunk 6 above)
Comment (mnesbit) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: Sadly filterIndexed returns a List<T>, so I would lose the index needed for the StateRef

**Comment by mnesbit** [comment]
I tried using filterIsInstance and it just doesn't work for us and in fact leads to more uneccessary mapping then filtering, not less. I have switch to Class.isInstance as suggested though.

File: core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt [hunk 8]
```diff
@@ -56,4 +58,125 @@ abstract class BaseTransaction(
...
+     * Clazz must be an extension of [ContractState].
+     * @return the single [StateAndRef] item matching the predicate.
+     * @throws IllegalArgumentException if no item, or multiple items are found matching the requirements.
+     */
+    fun <T : ContractState> findOutRef(predicate: Predicate<T>, clazz: Class<T>): StateAndRef<T> {
+        return filterOutRefs(predicate, clazz).single()
+    }
+
+    /**
+     * Kotlin extension methods to take advantage of Kotlin's smart type inference when querying the LedgerTransaction
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/transactions/BaseTransaction.kt]: This comment is only applicable to this method so best to make it a `//` comment

**Pull request merged**